jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java

```diff
@@ -0,0 +1,384 @@
+        verticesDataMap = new HashMap<>();
+    }
+
+    /**
+     * Asserts that all edges in the {@link #graph} have positive weights.
+     */
+    private void assertPositiveWeights() {
+        boolean allEdgesWithNonNegativeWeights = graph.edgeSet().stream()
+                .map(graph::getEdgeWeight).allMatch(weight -> weight >= 0);
+        if (!allEdgesWithNonNegativeWeights) {
+            throw new IllegalArgumentException(NEGATIVE_EDGE_WEIGHT_NOT_ALLOWED);
+        }
+    }
+
+    /**
+     * Calculates max edge weight in the {@link #graph}.
+     *
+     * @return max edge weight
+     */
+    private Optional<Double> getMaxEdgeWeight() {
+        return graph.edgeSet().stream().map(graph::getEdgeWeight).max(Double::compare);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public GraphPath<V, E> getPath(V source, V sink) {
+        if (!graph.containsVertex(source)) {
+            throw new IllegalArgumentException(GRAPH_MUST_CONTAIN_THE_SOURCE_VERTEX);
+        }
+        if (!graph.containsVertex(sink)) {
+            throw new IllegalArgumentException(GRAPH_MUST_CONTAIN_THE_SINK_VERTEX);
+        }
+        return getPaths(source).getPath(sink);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public SingleSourcePaths<V, E> getPaths(V source) {
+        if (!graph.containsVertex(source)) {
+            throw new IllegalArgumentException(GRAPH_MUST_CONTAIN_THE_SOURCE_VERTEX);
+        }
+        assertPositiveWeights();
+
+        if (delta == 0.0) {
+            delta = findDelta();
+        }
+        fillMaps();
+        numOfBuckets = numOfBuckets();
+
+        computeShortestPaths(source);
+
+        Map<V, Pair<Double, E>> distanceAndPredecessorMap = new HashMap<>();
+        for (Map.Entry<V, AtomicReference<Triple<Integer, Double, E>>> entry : verticesDataMap.entrySet()) {
+            distanceAndPredecessorMap.put(entry.getKey(),
+                    Pair.of(entry.getValue().get().getSecond(),
+                            entry.getValue().get().getThird()));
+        }
+        return new TreeSingleSourcePathsImpl<>(graph, source, distanceAndPredecessorMap);
+    }
+
+    /**
+     * Calculates value of {@link #delta}. The value is calculated to
+     * maximal edge weight divided by maximal out-degree in the {@link #graph}
+     * or $1.0$ if edge set of the {@link #graph} is empty.
+     *
+     * @return bucket width
+     */
+    private double findDelta() {
+        Optional<Double> maxEdgeWeight = getMaxEdgeWeight();
+        if (maxEdgeWeight.isPresent()) {
+            int maxOutDegree = graph.vertexSet().stream().mapToInt(graph::outDegreeOf).max().orElse(0);
+            return maxEdgeWeight.get() / maxOutDegree;
+        } else {
+            return 1.0;
+        }
+    }
+
+    /**
+     * Fills {@link #light}, {@link #heavy} and{@link #verticesDataMap} fields.
+     */
+    private void fillMaps() {
+        graph.vertexSet().forEach(v -> {
```

Comment:
If you change the maps to `ConcurrentHashMap` can you perform this also in parallel?